src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -76,9 +87,78 @@ internal abstract class JsonPropertyInfo
         internal virtual void GetPolicies(JsonSerializerOptions options)
         {
             DetermineSerializationCapabilities(options);
+            DeterminePropertyName(options);
             IgnoreNullValues = options.IgnoreNullValues;
         }
 
+        private void DeterminePropertyName(JsonSerializerOptions options)
+        {
+            if (PropertyInfo != null)
+            {
+                JsonPropertyNameAttribute nameAttribute = GetAttribute<JsonPropertyNameAttribute>();
+                if (nameAttribute != null)
+                {
+                    NameAsString = nameAttribute.Name;
+                    
+                    // This is detected and thrown by caller.
+                    if (NameAsString == null)
+                    {
+                        return;
+                    }
+                }
+                else if (options.PropertyNamingPolicy != null)
+                {
+                    NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
+
+                    // This is detected and thrown by caller.
+                    if (NameAsString == null)
+                    {
+                        return;
+                    }
+                }
+                else
+                {
+                    NameAsString = PropertyInfo.Name;
+                }
+
+                // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
+                _name = Encoding.UTF8.GetBytes(NameAsString);
+
+                // Set the compare name.
+                if (options.PropertyNameCaseInsensitive)
+                {
+                    CompareNameAsString = NameAsString.ToUpperInvariant();
+                    _compareName = Encoding.UTF8.GetBytes(CompareNameAsString);
+                }
+                else
+                {
+                    CompareNameAsString = NameAsString;
+                    _compareName = _name;
+                }
+
+                // Cache the escaped name.
+                int valueIdx = JsonWriterHelper.NeedsEscaping(_name);
+                if (valueIdx == -1)
+                {
+                    _escapedName = _name;
+                }
+                else
+                {
+                    int length = JsonWriterHelper.GetMaxEscapedLength(_name.Length, valueIdx);
+
+                    byte[] tempArray = ArrayPool<byte>.Shared.Rent(length);
+
+                    JsonWriterHelper.EscapeString(_name, tempArray, valueIdx, out int written);
+                    _escapedName = new byte[written];
```

Comment:
This is unfortunate, but I see the need. Any way to avoid it? Maybe by storing a Memory<T> instead? We effectively want slicing here.